src/axolotl/core/trainer_builder.py

```diff
@@ -963,9 +963,25 @@ def tokenize_row(
         self, feature, model: Optional[Union[PreTrainedModel, torch.nn.Module]] = None
     ) -> Dict:
         res = super().tokenize_row(feature, model=model)
+        # fix when the tokenizer doesn't have a bos_token_id, e.g. Qwen
         if self.tokenizer.bos_token_id is None and res["prompt_input_ids"][0] is None:
             for key in res.keys():
                 res[key] = res[key][1:]
+
+        if self.tokenizer.bos_token_id is not None:
+            # dpo trainer is incorrectly prepending the bos_token_id to the dpo outputs
+            if res["chosen_input_ids"][0] == self.tokenizer.bos_token_id:
+                res["chosen_input_ids"] = res["chosen_input_ids"][1:]
+                res["chosen_labels"] = res["chosen_labels"][1:]
+                res["chosen_attention_mask"] = res["chosen_attention_mask"][1:]
+            if res["rejected_input_ids"][0] == self.tokenizer.bos_token_id:
+                res["rejected_input_ids"] = res["rejected_input_ids"][1:]
+                res["rejected_labels"] = res["rejected_labels"][1:]
+                res["rejected_attention_mask"] = res["rejected_attention_mask"][1:]
+            if res["prompt_input_ids"][0] == self.tokenizer.bos_token_id:
+                res["prompt_input_ids"] = res["prompt_input_ids"][1:]
+                res["prompt_attention_mask"] = res["prompt_attention_mask"][1:]
+
```

Comment:
From the linked Issue, we should only remove the BOS from the `chosen`/`rejected`, right? The `prompt` should not require it?

If it does require removing bos, does `res["prompt_labels"]` exist & do we need to remove it also?